crates/adapters/src/controller/stats.rs

```diff
@@ -51,18 +53,67 @@ use metrics::{KeyName, SharedString as MetricString, Unit as MetricUnit};
 use metrics_util::{debugging::DebugValue, CompositeKey};
 use ordered_float::OrderedFloat;
 use rand::{seq::index::sample, thread_rng};
-use serde::{ser::SerializeStruct, Serialize, Serializer};
+use serde::{ser::SerializeStruct, Deserialize, Serialize, Serializer};
 use std::{
     cmp::min,
-    collections::{BTreeMap, BTreeSet},
+    collections::{BTreeMap, BTreeSet, VecDeque},
     sync::{
         atomic::{AtomicBool, AtomicU64, Ordering},
         Arc, Mutex,
     },
 };
-use tracing::{debug, error, info};
+use tracing::{debug, error, info, warn};
 use uuid::Uuid;
 
+use utoipa::ToSchema;
+
+/// Completion token.
+///
+/// A completion token associated with an endpoint identifies a position in the endpoint's
+/// input stream and can be used to check whether all inputs up to this position have been
+/// fully processed by the pipeline, including output changes getting processed by all
+/// output connectors.
+#[derive(Debug, Serialize, Deserialize, ToSchema)]
+pub struct CompletionToken {
+    /// Pipeline incarnation uuid. Used to reject tokens generated by previous incarnations
+    /// of the pipeline (before suspend, crash, or shutdown).
+    #[serde(rename = "u")]
+    incarnation: Uuid,
```

Comment:
so if a pipeline restarts you can't find out anything about what is completed?

Reply:
No. In theory we could do it with exactly-once fault tolerance by journaling token list, but we don't do it at the moment.